### DISCUSSION TITLE
Can't get updated metadata to embed

### DISCUSSION THREAD

**Discussion by fsumarc** [advplyr/audiobookshelf#1845] [Q&A] [answered]
Hi! I apologize if this is a simple issue. I'm new to docker and such.  Not new to networking and servers, I have spent many hours searching and digging for answers. Sorry if it ends up being a wall of text. I want to try to explain what all I have done so far. I have a windows 11 machine. I  have Plex for my video and music.  I had my audiobooks in there as well but was a clunky workaround that was more frustrating that enjoyable to use it.  So, I did some digging and found Audiobookshelf.  It looked like exactly what I needed.  I followed a tutorial online for setting up Docker Desktop and Portainers.  Setup the Audiobookshelf container. Everything seemed to be going well. I even set up Cloudflare so I can access my stuff from everywhere. It's great!  

Here's the issue.   My audiobooks are collected over several years and many sources.  The metadata tags are a mess.  Several different formats.  I originally pulled in a chunk of books with the directories formatted as recommended.  Scan and the books showed up.  I went through them and matched them up, cleaned all the metadata, converted to mb4, rechecked metadata, and then ran "embed metadata in files". Everything seemed to be going great.  I pulled in chunks at a time to clean them up and spent a couple weeks using my free time cleaning and organizing. I am right around 400 books so far.   Many more to go.   

I noticed while working on batches that sometimes it would take scan and restart container to get books added to show up.  Sometimes it created error duplicates and such. But once they did I could clean them up. I seemed to have to refix "series" stuff often though.  Sometimes titles would be messed up again, etc.  I decided to do a force rescan to see what's up and noticed a lot of stuff messed up again.  I realized it had reset the metadata to what the files originally had.  I brought up mp3tag and configured the set of tags in it to read what ABS used and realized it had never updated any of my metadata.  It just all looked good from the database. 

I started testing some stuff. Open the file in mp3tag and see messed up data. Go into ABS and do match and reset chapters and all that so everything was nice. Run embed metadata in ABS.  Check ABS logs. They say completed embed.  Go back out to the file in mp3tag and refresh.   Still old data.  WTH?!   So I did a test with a series of books and fixed all the metadata for them in mp3tag. Forced rescan in ABS and they loaded in fine and come up correctly.  

My first thought was permission issues. Though the logs saying it completed the embed successfully had me confused.  I think there is a bug there.  I initially set up v2.2.20. I saw that there was an update to 2.2.21 so I ran that process.  Seemed to go fine but didn't change any of the issues.  I went through windows security permissions for the audiobook library and all seems ok.   I checked my antivirus and there is no mention of anything to do with ABS or Docker.

So, my current though is that there is an issue with the app in container not being able to edit files in windows?  It seems to work fine converting mp3 and such to m4b but can't edit existing files.  An interesting thing I noticed is that even though it doesn't seem to embed the updated data windows updates the "modified date" to the exact time and date on the file as when it tried to edit the data. I assume ABS is able to open and resave but not change.   Help!   I don't know where to go from here.  I really dont want to have to manually edit all metadata externally.   ABS is exactly what I am looking for. I just need it to work correctly.  I'm sure it's something I've done but I have no clue where to go now.

Side note, I also keep fixing author data and some of  those keep rebreaking as well. Not sure if related.

**Comment by advplyr** [comment]
Sometimes when you are changing the names of folders in your filesystem Abs will mistakenly think this is a new book and create the duplicate. The old book will then get marked as missing after a re-scan. This is something I'm aware of and should be able to improve.

What is helpful in these situations is to try to find the smallest number of steps so someone can reproduce the issue.
Here is what I just tested on Windows 10 using internal hard drives.

1. Setup a test book with a single mp3 audio file and noted the meta tags shown in file explorer
2. In Abs I changed the title from "Tester" to "Tester2" and saved
3. Pressed edit -> Tools -> Quick Embed
4. Checked file explorer to see "Tester2" is now shown

![image](https://github.com/advplyr/audiobookshelf/assets/67830747/da4952bb-f4db-4512-87e6-221db4826f91)

**Comment by fsumarc** [comment]
Thank you for the reply!  That is basically what I was doing except in Windows 11.  That process seems to work some of the time but for a bunch of the books it just wont update.  I can edit the metadata externally in mp3tag and if I rescan ABS updates to new data.  But they wont update the other direction.  I wish it was all files then I could track it down better. I am doing some testing now on different types of files and processes to see if I can find a cause.  All I know is if a file won't embed the first time, nothing I can do will make it work.   I'll post more info as I find more details.   Lemme know if anything specific I should try.

**Comment by fsumarc** [comment, marked as answer]
Ok if anyone ends up here looking into same issue I think I found my solution.  My audiobook collection comes from many years and many places so are in many formats.  I noticed that every book that had the issue with not updating metadata were all m4b format. Digging into files that were not working appears to be ones that are in weird formats like mov, mpg, iso, etc embedded n the m4b wrapper.  As opposed to mp3.  Each problem file I converted from m4b to mp3 and could immediately get metadata to save correctly via ABS.  Even after converting back to m4b they work fine and can even be edited again.
